DESIGN.md

```diff
@@ -0,0 +1,211 @@
+
+- *pkg*
+    - *ipc*: types and methods required for inter-process communication.
+    - *log*: implement a common logging interface accross all components. This will not be developed from scratch,
+    just adapt the logger used in other projects
+    - *alps*: library code for creating and listening to Arbitrary Linux Process Service.
+- *cmd*
+    - *api*: HTTP API server using mTLS authentication with ACL authorization. Will communicate with the job 
+    worker via RPC. Authentication and authorization will be done via middleware. Any endpoint will return 
+    `{"status": 401, "error": "you are not authorized to perform that action"}` if the request does not have a recognized
+    user.
+        - **POST /start** starts a process and returns an ID to use for getting output and control
+            - Request: 
+            ```json
+            {"job": "ls -lh /var/log"}
+            ```
+            - Success response: 
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z" 
+            }
+            ``` 
+            - Error response: 
+            ```json
+            {"status": 500, "error": "some failure"}
+            ```
+        - **POST /stop/:jobID** stops the job with the given ID
+            - Success response: 
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z", 
+                "stopTime": "2020-08-18T12:00:00Z" ,
+                "exitCode": 0
+            }
+            ```
+            - Error response 
+            ```json
+            {"status": 500, "error": "some failure"}
+            ```
+        - **GET /status/:jobID** returns the status of the job with the given ID
+            - Success response:
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
+                "job": "ls -lh /var/log",
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z", 
+                "jobStatus": "stopped",
+                "stopTime": "2020-08-18T12:00:00Z" ,
+                "exitCode": 0
+            }
+            ```
+            - Error response: 
+            ```json
+            {"status": 404, "error": "no job found with that ID"}
+            ```
+        - **GET /output/:jobID** returns the output of a job
+            - Success response:
+            ```json
+            {
+                "status": 200, 
+                "jobID": 123, 
+                "job": "ls -lh /var/log",
+                "jobStatus": "stopped",
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z", 
+                "stopTime": "2020-08-18T12:00:00Z" ,
+                "exitCode": 0
+                "stderrSize": 123,
```

Comment:
Why include the size?

Reply:
No reason besides convenience for the caller.